polars/polars-core/src/datatypes/dtype.rs

```diff
@@ -71,7 +71,12 @@ impl PartialEq for DataType {
                 // Don't include rev maps in comparisons
                 #[cfg(feature = "dtype-categorical")]
                 (Categorical(_), Categorical(_)) => true,
-                (Datetime(tu_l, tz_l), Datetime(tu_r, tz_r)) => tu_l == tu_r && tz_l == tz_r,
+                (Datetime(tu_l, tz_l), Datetime(tu_r, tz_r)) => {
```

Comment:
I think we need to do the equality overwrite for the wildcard case a level higher.


You now have a situation where the datetime can be equal where the hash isn't and that's not good. We can make a special comparison function for `DataType` and call that one on the location where we trigger the selection API. (I am not sure where we do, but I can figure that out for you).

Reply:
Sure, point me at it ;)
(About to hop on a train, so can take care of it tonight).

Reply:
This `dtype_contains` needs to be replaced with something like:

```rust
    dtypes.iter().any(|dtype| special_eq(dtype, f.dtype));
```

where `special_eq` has a special branch for `Datetime("*")`.

Reply:
Thanks :)  Will dive in when I get back tonight.

Reply:
@ritchie46: Done. Spotted that we also needed to apply it in the exclusion-related func below (and added a bonus unit test for coverage). While I was at it I overhauled selector inversion; since you added a proper `sub` set op it's as simple as `all() - selector` - this allowed for some extra simplification elsewhere :)